src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt

```diff
@@ -0,0 +1,369 @@
+                return passedQuery
+            }
+            throw OptimizedQueryException()
+        }
+
+        private fun withClauseWithOptionalDistinct(
+                exposesWith: StatementBuilder.ExposesWith,
+                withs: Collection<Expression>,
+                useDistinct: Boolean = true): StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere =
+                when {
+                    useDistinct && withs.size == 1 -> exposesWith.withDistinct(*withs.toTypedArray())
+                    else -> exposesWith.with(*withs.toTypedArray())
+                }
+
+        private fun <WithWhere> addConditions(
+                propertyContainer: PropertyContainer<*>,
+                variablePrefix: String,
+                conditions: List<Predicate<FieldOperator>>,
+                filterParams: MutableMap<String, Any?>,
+                conditionAdder: (where: WithWhere?, condition: Condition) -> WithWhere
+        ): WithWhere? {
+            var where: WithWhere? = null
+            for (conditionField in conditions) {
+                val (predicate, objectField, field) = conditionField
+                val prop = propertyContainer.property(field.name)
+                val parameter = Cypher.parameter(variablePrefix + "_" + objectField.name)
+                val condition = predicate.conditionCreator(prop, parameter)
+                filterParams[parameter.name] = objectField.value.toJavaValue()
+                where = conditionAdder(where, condition)
+            }
+            return where
+        }
+
+
+        private fun createRelation(rel: RelationshipInfo, start: Node, end: Node): Relationship =
+                when (rel.out) {
+                    false -> start.relationshipFrom(end, rel.relType)
+                    true -> start.relationshipTo(end, rel.relType)
+                    null -> start.relationshipBetween(end, rel.relType)
+                }
+    }
+
+    companion object {
+        private val LOGGER = LoggerFactory.getLogger(OptimizedFilterHandler::class.java)
+
+        data class Predicate<T>(
+                val op: T,
+                val queryField: ObjectField,
+                val fieldDefinition: GraphQLFieldDefinition,
+                val index: Int)
+
+        class ParsedQuery(objectValue: ObjectValue, type: GraphQLFieldsContainer) {
+            val fieldPredicates = mutableListOf<Predicate<FieldOperator>>()
+            val relationPredicates = mutableListOf<Predicate<RelationOperator>>()
+            val or: List<Value<*>>?
+            val and: List<Value<*>>?
+
+            init {
+                val queriedFields = objectValue.objectFields
+                    .mapIndexed { index, field -> field.name to (index to field) }
+                    .toMap(mutableMapOf())
+                or = queriedFields.remove("OR")?.second?.value?.let {
+                    return@let (it as ArrayValue).values
```

Comment:
not sure we need the return@let here. 
This should work too: `value?.let { if (false) "a" else null} ?: throw RuntimeException()`
there should be a shorter way with a conditional let but I don't have an IDE running currently.

Reply:
we cannot pull the `throws` out of the let since this would change the logic